sql/src/main/java/io/crate/execution/engine/collect/PKLookupOperation.java

```diff
@@ -177,60 +135,46 @@ public void runWithShardProjections(UUID jobId,
-        }
-        ArrayList<BatchIterator<Row>> iterators = new ArrayList<>(shardAndIdsList.size());
-        for (ShardAndIds shardAndIds : shardAndIdsList) {
-            Stream<Row> rowStream = shardAndIds.value.stream()
-                .map(pkAndVersion -> lookupDoc(shardAndIds.shard,
-                                               pkAndVersion.id(),
-                                               pkAndVersion.version(),
-                                               pkAndVersion.seqNo(),
-                                               pkAndVersion.primaryTerm()))
+            Stream<Row> rowStream = idsByShardEntry.getValue().stream()
+                .map(pkAndVersion -> lookupDoc(
+                    shard,
+                    pkAndVersion.id(),
+                    pkAndVersion.version(),
+                    VersionType.EXTERNAL,
+                    pkAndVersion.seqNo(),
+                    pkAndVersion.primaryTerm()))
+                .filter(Objects::nonNull)
                 .map(resultToRow);
 
-            Projectors projectors = new Projectors(
-                projections,
-                jobId,
-                txnCtx,
-                ramAccountingSupplier.get(),
-                memoryManagerSupplier.get(),
-                shardAndIds.projectorFactory);
-            final Iterable<Row> rowIterable;
-            if (nodeConsumer.requiresScroll() && !projectors.providesIndependentScroll()) {
-                rowIterable = rowStream.map(row -> new RowN(row.materialize())).collect(Collectors.toList());
+            if (projections.isEmpty()) {
+                final Iterable<Row> rowIterable = requiresScroll
+                    ? rowStream.map(row -> new RowN(row.materialize())).collect(Collectors.toList())
+                    : rowStream::iterator;
+                iterators.add(InMemoryBatchIterator.of(rowIterable, SentinelRow.SENTINEL, true));
             } else {
-                rowIterable = rowStream::iterator;
+                ProjectorFactory projectorFactory;
+                try {
+                    projectorFactory = shardCollectSource.getProjectorFactory(shardId);
+                } catch (ShardNotFoundException e) {
+                    if (ignoreMissing) {
+                        continue;
+                    }
+                    throw e;
+                }
+                Projectors projectors = new Projectors(
+                    projections,
+                    jobId,
+                    txnCtx,
+                    ramAccountingSupplier.get(),
+                    memoryManagerSupplier.get(),
+                    projectorFactory);
+                final Iterable<Row> rowIterable = requiresScroll && !projectors.providesIndependentScroll()
+                    ? rowStream.map(row -> new RowN(row.materialize())).collect(Collectors.toList())
+                    : rowStream::iterator;
+                iterators.add(InMemoryBatchIterator.of(rowIterable, SentinelRow.SENTINEL, true));
```

Comment:
isn't `projectors.wrap(...)` missing here?

Reply:
Right. Too much back & forth. I'll also append some changes that this code is actually used again.